### PR TITLE
8253725: Jextract fails to extract big monolithic headers

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch adds support for splitting the symbols in the main header class into multiple classes; the header class is then rewrited to extend all these additional classes, so that all symbols would be exposed to users in the classic way.

I've toyed with some ways to split on a by-header basis, but found that in certain cases we would generate near-empty classes - and that would still not prevent issues from occurring in the corner case a single very big header file contained so many symbols.

I've reorganized the code a bit - I noted that we were generating several nested classes, but in some cases (structs) we were using a builder, in other cases (functional interfaces, typedef) we weren't. Now we use builders for classes, and I've added an intermediate nested builder abstraction which does the right thing for all nested cases.

I've also converted the source constant generator to be a JavaSourceBuilder too - since it too is a class, after all, with a classBegin/End and build methods.

Finally, I moved the string builder/indentation logic out to a different class (StringSourceBuilder); since some builders want to have their own string buffer, while others (the nested ones) don't.

The organization of the builder code is still far from optimal, but I think that, moving forward, this consolidation should make it easier to make changes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253725](https://bugs.openjdk.java.net/browse/JDK-8253725): Jextract fails to extract big monolithic headers

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/357/head:pull/357`
`$ git checkout pull/357`

### CODE REVIEW

**Review by JornVernee — APPROVED** [review]
Looks great overall! I left some comments inline.

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java [hunk 1]
```diff
@@ -309,8 +305,8 @@ public Void visitFunction(Declaration.Function funcTree, Declaration parent) {
                     warn("varargs in callbacks is not supported");
                 }
                 MethodType fitype = typeTranslator.getMethodType(f, false);
-                String anno = annotationWriter.getCAnnotation(param.type());
-                toplevelBuilder.addFunctionalInterface(name, fitype, Type.descriptorFor(f).orElseThrow(), anno);
+                toplevelBuilder.addFunctionalInterface(name, fitype,
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java:308]: No - this one of the (many) tricky things. Struct/Functional interface/Typedef nested classes always end up in the toplevel (the API the user can see).

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java [hunk 2]
```diff
@@ -357,7 +353,7 @@ public Void visitTypedef(Declaration.Typedef tree, Declaration parent) {
                              * };
                              */
                             if (structDefinitionSeen(s)) {
-                                toplevelBuilder.emitTypedef(tree, structDefinitionName(s));
+                                toplevelBuilder.addTypeDef(tree.name(), structDefinitionName(s), tree.type());
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java:356]: Same as above

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/SourceConstantHelper.java [hunk 3]
```diff
@@ -228,65 +209,28 @@ private DirectMethodHandleDesc getGetterDesc(String name, Class<?> type) {
...
+    @Override
+    protected void addImportSection() {
+        builder.append("import java.lang.invoke.MethodHandle;\n");
+        builder.append("import java.lang.invoke.VarHandle;\n");
+        builder.append("import jdk.incubator.foreign.*;\n");
+        builder.append("import jdk.incubator.foreign.MemoryLayout.PathElement;\n");
+        builder.append("import static ");
+        builder.append(OutputFactory.C_LANG_CONSTANTS_HOLDER);
+        builder.append(".*;\n\n");
     }
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/SourceConstantHelper.java:234]: Likely forgot to remove, I'll try and see.

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        files.addAll(headers.stream()
+                .flatMap(hf -> hf.build().stream())
+                .collect(Collectors.toList()));
+        return files;
+    }
+
+    void addFunctionalInterface(String name, MethodType mtype, FunctionDescriptor desc, Type type) {
+        FunctionalInterfaceBuilder builder = new FunctionalInterfaceBuilder(this, name, mtype, desc, type);
+        builder.classBegin();
+        builder.classEnd();
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:75]: I see what you mean. Right now, one piece of messiness is that the class contents are attached to either class begin/end, w/o much discipline. I'll see what I can do.

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            HeaderFileBuilder headerFileBuilder = new HeaderFileBuilder(className + "$" + headers.size(), pkgName,
+                    lastHeader().map(h -> h.className).orElse(null),
+                    constantHelper, annotationWriter);
+            headerFileBuilder.classBegin();
+            headers.add(headerFileBuilder);
+            declCount = 1;
+            return headerFileBuilder;
+        } else {
+            declCount++;
+            return lastHeader().orElse(this);
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:103]: yes - that leads to a CP fill of about half - we could probably increase it further, but seems a slippery slope

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:75]: I did some investigation here - and it doesn't seem like adding extra structure here is worth it. Adding build might risk (if we make it general) to conflict with HeaderBuilder.build which does a different task.

As for `addStuff` while in principle that's obvious, in practice that's a bundle of static code that we want to generate with the class - so if we go for that we're asking all clients to do classBegin/addStuff/classEnd (OutputFactory would be affected badly).

So, yes, as is it looks awful, but I think we should try to fix it in a more radical rewriting of this code.

[the remark this answers is not in the document]
Comment (JornVernee) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:75]: Ok, sounds good

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**